quiche/src/lib.rs

```diff
@@ -8388,13 +8554,144 @@ impl TransportParams {
+        addr_v4: Option<SocketAddrV4>, addr_v6: Option<SocketAddrV6>,
+        connection_id: ConnectionId<'static>, stateless_reset_token: u128,
+    ) -> Self {
+        PreferredAddressParams {
+            addr_v4,
+            addr_v6,
+            connection_id,
+            stateless_reset_token,
+        }
+    }
+
+    fn decode(b: &mut octets::Octets) -> Result<Self> {
+        let addr_v4 = PreferredAddressParams::decode_v4(b)?;
+        let addr_v6 = PreferredAddressParams::decode_v6(b)?;
+
+        if addr_v4.is_none() && addr_v6.is_none() {
+            return Err(Error::InvalidTransportParam);
+        }
+
+        let cid = b.get_bytes_with_varint_length()?.to_vec().into();
+        let stateless_reset_token = u128::from_be_bytes(
+            b.get_bytes(16)?.to_vec().try_into().unwrap(), /* Since a 128 bit
+                                                            * vector has just
+                                                            * successfully
+                                                            * been created
+                                                            * above, this
+                                                            * should never
+                                                            * panic. */
+        );
+
+        Ok(PreferredAddressParams::new(
+            addr_v4,
+            addr_v6,
+            cid,
+            stateless_reset_token,
+        ))
+    }
+
+    fn decode_v4(b: &mut octets::Octets) -> Result<Option<SocketAddrV4>> {
+        let mut addr_v4 = None;
+        let ip_v4_b = <[u8; 4]>::try_from(b.slice(4)?)
+            .map_err(|_| Error::InvalidTransportParam)?;
+        b.skip(4)?;
+        let ip_v4 = Ipv4Addr::from(ip_v4_b);
+
+        let port_v4 = b.get_u16()?;
+
+        if ip_v4.is_unspecified() ^ (port_v4 == 0) {
+            return Err(Error::InvalidTransportParam);
+        }
+
+        if !ip_v4.is_unspecified() {
+            addr_v4 = Some(SocketAddrV4::new(ip_v4, port_v4));
+        }
+
+        Ok(addr_v4)
+    }
+
+    fn decode_v6(b: &mut octets::Octets) -> Result<Option<SocketAddrV6>> {
+        let mut addr_v6 = None;
+        let ip_v6_b = <[u8; 16]>::try_from(b.slice(16)?)
+            .map_err(|_| Error::InvalidTransportParam)?;
+        b.skip(16)?;
+        let ip_v6 = Ipv6Addr::from(ip_v6_b);
+        let port_v6 = b.get_u16()?;
+
+        if ip_v6.is_unspecified() ^ (port_v6 == 0) {
+            return Err(Error::InvalidTransportParam);
+        }
+
+        if !ip_v6.is_unspecified() {
+            addr_v6 = Some(SocketAddrV6::new(ip_v6, port_v6, 0, 0));
+        }
+
+        Ok(addr_v6)
+    }
+
+    /// Encode the preferred address parameters into a buffer.
+    ///
+    /// # Arguments
+    ///
+    /// * `preferred_address_params` - A reference to the preferred address
+    ///   parameters to be encoded.
+    /// * `buffer` - A reference to a buffer to encode the
+    ///   `preferred_address_params` into.
```

Comment:
This comment doesn't seem particularly useful, especially since it's internal code

Reply:
removed